ligo/token/fa2.mligo

```diff
@@ -0,0 +1,147 @@
+// SPDX-FileCopyrightText: 2021 Arthur Breitman
+// SPDX-License-Identifier: LicenseRef-MIT-Arthur-Breitman
+
+#include "../types.mligo"
+
+// -----------------------------------------------------------------
+// Helper
+// -----------------------------------------------------------------
+
+[@inline]
+let check_position_exist (position_id, store : position_id * storage) : position_id =
+  let valid_position_id =
+      match (Big_map.find_opt position_id store.positions) with
+        | Some _ -> position_id
+        | None -> ([%Michelson ({| { FAILWITH } |} : string * unit -> position_id)]
+              ("FA2_TOKEN_UNDEFINED", ()) : position_id)
+  in valid_position_id
+
+[@inline]
+let check_sender (from_ , store : address * storage): address =
+  if (sender = from_) then from_
+  else
+    let key: operator = { owner = from_; operator = sender} in
+    if Big_map.mem key store.operators then
+      from_
+    else
+     ([%Michelson ({| { FAILWITH } |} : string * unit -> address)]
+        ("FA2_NOT_OPERATOR", ()) : address)
+
+(* A function that combines the usual FA2's `debit_from` and `credit_to`.
+If `amt` is not equal to `0` or `1`, it will cause an error.
+The `position_id` is expected to exist. *)
+[@inline]
+let change_position_owner (amt, position_id, from_, to_, store: nat * position_id * address * address * storage): (ledger * position_by_index_map) =
+  if amt = 0n then
+    (store.ledger, store.positions_by_index) // We allow 0 transfer
+  else
+    let ledger = store.ledger in
+    let positions_by_index = store.positions_by_index in
+
+    let old_from_ledger_val = match Big_map.find_opt (from_, position_id) ledger with
+          Some ledger_val -> ledger_val
+        | None -> ([%Michelson ({| { FAILWITH } |} : string * (nat * nat) -> ledger_value)]
+              ("FA2_INSUFFICIENT_BALANCE", (amt, 0n)) : ledger_value) in
+
+    // We remove `from` entry from ledger.
+    let updated_from_ledger_val = match Michelson.is_nat (old_from_ledger_val.bal - amt) with
+          Some new_bal ->
+            if (new_bal <> 0n) then (failwith internal_impossible_err : ledger_value option)
```

Comment:
Here and slightly below we rely on the fact that each `token_id` is NFT, but this can also go wrong due to a bug or incorrect deployment. I would also use dedicated error(s) since we already went with multiple internal errors.

Reply:
I've added these errors to cover these cases, let me know if this is good enough:

```
| 322 | `internal_nft_amt_not_one` | Expect the position amount to be exactly one. |
| 323 | `internal_nft_belong_to_multi_owners` | Expect a position to belong to only one owner at a time. |
```